File: go/pkg/dns/dns_test.go
package dns

import (
	"strings"
	"testing"

	"github.com/ThirdKeyAi/agentpin/go/pkg/crypto"
	"github.com/ThirdKeyAi/agentpin/go/pkg/jwk"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func discoveryDoc(jwks []types.JWK) *types.DiscoveryDocument {
	return &types.DiscoveryDocument{
		AgentpinVersion:    "0.3",
		Entity:             "example.com",
		EntityType:         types.EntityMaker,
		PublicKeys:         jwks,
		MaxDelegationDepth: 0,
		UpdatedAt:          "2026-05-01T00:00:00Z",
	}
}

func TestParseFullRecord(t *testing.T) {
	r, err := ParseTxtRecord("v=agentpin1; kid=acme-2026-04; fp=sha256:abcd1234")
	if err != nil {
		t.Fatal(err)
	}
	if r.Version != "agentpin1" || r.Kid != "acme-2026-04" || r.Fingerprint != "sha256:abcd1234" {
		t.Fatalf("unexpected: %+v", r)
	}
}

func TestParseMinimalRecord(t *testing.T) {
	r, err := ParseTxtRecord("v=agentpin1;fp=sha256:abc")
	if err != nil {
		t.Fatal(err)
	}
	if r.Kid != "" || r.Fingerprint != "sha256:abc" {
		t.Fatalf("unexpected: %+v", r)
	}
}

func TestParseLowercasesFingerprint(t *testing.T) {
	r, err := ParseTxtRecord("v=agentpin1; fp=SHA256:ABCDEF")
	if err != nil {
		t.Fatal(err)
	}
	if r.Fingerprint != "sha256:abcdef" {
		t.Fatalf("expected lowercased fp, got %q", r.Fingerprint)
	}
}

func TestParseToleratesWhitespaceAndOrder(t *testing.T) {
	r, err := ParseTxtRecord("  fp = sha256:beef ;  v = agentpin1  ")
	if err != nil {
		t.Fatal(err)
	}
	if r.Fingerprint != "sha256:beef" || r.Version != "agentpin1" {
		t.Fatalf("unexpected: %+v", r)
	}
}

func TestParseIgnoresUnknownFields(t *testing.T) {
	r, err := ParseTxtRecord("v=agentpin1; fp=sha256:abc; future=ignoreme")
	if err != nil {
		t.Fatal(err)
	}
	if r.Fingerprint != "sha256:abc" {
		t.Fatalf("unexpected: %+v", r)
	}
}

func TestParseFailures(t *testing.T) {
	cases := []string{
		"fp=sha256:abc",               // missing v
		"v=agentpin1",                 // missing fp
		"v=agentpin99; fp=sha256:abc", // bad version
		"v=agentpin1; fp=abc",         // bad fp prefix
		"v=agentpin1; broken",         // missing '='
		"v=schemapin1; fp=sha256:abc", // SchemaPin format
	}
	for _, c := range cases {
		if _, err := ParseTxtRecord(c); err == nil {
			t.Fatalf("expected error for %q", c)
		}
	}
}

func fpFor(j types.JWK) string {
	t := strings.ToLower(jwk.JWKThumbprint(&j))
	if !strings.HasPrefix(t, "sha256:") {
		t = "sha256:" + t
	}
	return t
}

func TestVerifyMatchAgainstSingleKey(t *testing.T) {
	kp, err := crypto.GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	j, err := jwk.PEMToJWK(kp.PublicKeyPEM, "kid-1")
	if err != nil {
		t.Fatal(err)
	}
	if err := VerifyDnsMatch(discoveryDoc([]types.JWK{j}),
		&TxtRecord{Fingerprint: fpFor(j)}); err != nil {
		t.Fatal(err)
	}
}

func TestVerifyMatchAgainstOneOfMultipleKeys(t *testing.T) {
	kp1, _ := crypto.GenerateKeyPair()
	kp2, _ := crypto.GenerateKeyPair()
	j1, _ := jwk.PEMToJWK(kp1.PublicKeyPEM, "kid-a")
	j2, _ := jwk.PEMToJWK(kp2.PublicKeyPEM, "kid-b")
	doc := discoveryDoc([]types.JWK{j1, j2})
	if err := VerifyDnsMatch(doc, &TxtRecord{Kid: "kid-b", Fingerprint: fpFor(j2)}); err != nil {
		t.Fatal(err)
	}
}

func TestVerifyKidMismatchFailsEvenWhenFpMatches(t *testing.T) {
	kp, _ := crypto.GenerateKeyPair()
	j, _ := jwk.PEMToJWK(kp.PublicKeyPEM, "kid-real")
	doc := discoveryDoc([]types.JWK{j})
	if err := VerifyDnsMatch(doc, &TxtRecord{Kid: "kid-different", Fingerprint: fpFor(j)}); err == nil {
		t.Fatal("expected mismatch error")
	}
}

func TestVerifyFingerprintMismatch(t *testing.T) {
	kp, _ := crypto.GenerateKeyPair()
	j, _ := jwk.PEMToJWK(kp.PublicKeyPEM, "kid-1")
	doc := discoveryDoc([]types.JWK{j})
	if err := VerifyDnsMatch(doc, &TxtRecord{
		Fingerprint: "sha256:0000000000000000000000000000000000000000000000000000000000000000",
	}); err == nil {
		t.Fatal("expected mismatch error")
	}
}

func TestTxtRecordNameStripsTrailingDot(t *testing.T) {
	if got := TxtRecordName("example.com"); got != "_agentpin.example.com" {
		t.Fatalf("got %q", got)
	}
	if got := TxtRecordName("example.com."); got != "_agentpin.example.com" {
		t.Fatalf("got %q", got)
	}
}

File: javascript/tests/dns.test.js
/**
 * Tests for DNS TXT cross-verification (v0.3.0).
 */

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

import {
    generateKeyPair,
    pemToJwk,
    jwkThumbprint,
    parseTxtRecord,
    verifyDnsMatch,
    txtRecordName,
    EntityType,
} from '../src/index.js';

function makeDiscovery(jwks) {
    return {
        agentpin_version: '0.3',
        entity: 'example.com',
        entity_type: EntityType.MAKER,
        public_keys: jwks,
        agents: [],
        max_delegation_depth: 0,
        updated_at: '2026-05-01T00:00:00Z',
    };
}

describe('parseTxtRecord', () => {
    it('parses a full record', () => {
        const r = parseTxtRecord('v=agentpin1; kid=acme-2026-04; fp=sha256:abcd1234');
        assert.equal(r.version, 'agentpin1');
        assert.equal(r.kid, 'acme-2026-04');
        assert.equal(r.fingerprint, 'sha256:abcd1234');
    });

    it('parses a minimal record', () => {
        const r = parseTxtRecord('v=agentpin1;fp=sha256:abc');
        assert.equal(r.version, 'agentpin1');
        assert.equal(r.kid, null);
        assert.equal(r.fingerprint, 'sha256:abc');
    });

    it('lowercases the fingerprint', () => {
        const r = parseTxtRecord('v=agentpin1; fp=SHA256:ABCDEF');
        assert.equal(r.fingerprint, 'sha256:abcdef');
    });

    it('tolerates whitespace and unsorted fields', () => {
        const r = parseTxtRecord('  fp = sha256:beef ;  v = agentpin1  ');
        assert.equal(r.version, 'agentpin1');
        assert.equal(r.fingerprint, 'sha256:beef');
    });

    it('ignores unknown fields for forward compat', () => {
        const r = parseTxtRecord('v=agentpin1; fp=sha256:abc; future=ignoreme');
        assert.equal(r.fingerprint, 'sha256:abc');
    });

    it('rejects missing v', () => {
        assert.throws(() => parseTxtRecord('fp=sha256:abc'));
    });

    it('rejects missing fp', () => {
        assert.throws(() => parseTxtRecord('v=agentpin1'));
    });

    it('rejects unsupported version', () => {
        assert.throws(() => parseTxtRecord('v=agentpin99; fp=sha256:abc'));
    });

    it('rejects fingerprint without sha256: prefix', () => {
        assert.throws(() => parseTxtRecord('v=agentpin1; fp=abc'));
    });

    it('rejects field without =', () => {
        assert.throws(() => parseTxtRecord('v=agentpin1; broken'));
    });

    it('rejects SchemaPin-format records', () => {
        // Sanity check: must not accidentally accept SchemaPin's TXT format.
        assert.throws(() => parseTxtRecord('v=schemapin1; fp=sha256:abc'));
    });
});

describe('verifyDnsMatch', () => {
    function fingerprintForKey(jwk) {
        const t = jwkThumbprint(jwk).toLowerCase();
        return t.startsWith('sha256:') ? t : `sha256:${t}`;
    }

    it('matches when TXT fp equals the key thumbprint', () => {
        const { publicKeyPem } = generateKeyPair();
        const jwk = pemToJwk(publicKeyPem, 'kid-1');
        const doc = makeDiscovery([jwk]);
        const txt = { kid: null, fingerprint: fingerprintForKey(jwk) };
        verifyDnsMatch(doc, txt);
    });

    it('matches one of multiple discovery keys', () => {
        const { publicKeyPem: pk1 } = generateKeyPair();
        const { publicKeyPem: pk2 } = generateKeyPair();
        const jwk1 = pemToJwk(pk1, 'kid-a');
        const jwk2 = pemToJwk(pk2, 'kid-b');
        const doc = makeDiscovery([jwk1, jwk2]);
        const txt = { kid: 'kid-b', fingerprint: fingerprintForKey(jwk2) };
        verifyDnsMatch(doc, txt);
    });

    it('fails when kid in TXT does not match the discovery key', () => {
        const { publicKeyPem } = generateKeyPair();
        const jwk = pemToJwk(publicKeyPem, 'kid-real');
        const doc = makeDiscovery([jwk]);
        const txt = { kid: 'kid-different', fingerprint: fingerprintForKey(jwk) };
        assert.throws(() => verifyDnsMatch(doc, txt));
    });

    it('fails on fingerprint mismatch', () => {
        const { publicKeyPem } = generateKeyPair();
        const jwk = pemToJwk(publicKeyPem, 'kid-1');
        const doc = makeDiscovery([jwk]);
        const txt = {
            kid: null,
            fingerprint: 'sha256:0000000000000000000000000000000000000000000000000000000000000000',
        };
        assert.throws(() => verifyDnsMatch(doc, txt));
    });
});

describe('txtRecordName', () => {
    it('strips trailing dot', () => {
        assert.equal(txtRecordName('example.com'), '_agentpin.example.com');
        assert.equal(txtRecordName('example.com.'), '_agentpin.example.com');
    });
});
